Support for Open Font Format kerning tables

One of the things I was working on, way back when, was adding support for Open Font Format kerning.  This is a form of kern table that is used by nearly all commercial fonts, but few free fonts.  Heirloom and n-t-roff do not know about this table, which results in the kerning values being read incorrectly from the font file.  (The OFF spec was released around the time Gunnar was winding down with Heirloom, so it is not at all surprising that there is no support for this table.)  The modifications were essentially complete back in 2018--the only things needed were to remove the debugging and trace messages--but that was in the same time frame I was running away from Github as fast as I could, and I abandoned it.  I ran across the files this past weekend while going through an old unused drive, and decided that they weren't doing anybody any good just setting there.

[1-otftest-before.pdf](https://github.com/n-t-roff/heirloom-doctools/files/14656274/1-otftest-before.pdf)

[2-otftest-after.pdf](https://github.com/n-t-roff/heirloom-doctools/files/14656275/2-otftest-after.pdf)

[testfiles.zip](https://github.com/n-t-roff/heirloom-doctools/files/14656276/testfiles.zip)

The two PDFs are "before" and "after" text samples showing the effect that using the OFF-style kern table can have.  The effect is generally subtle (the text looks less jittery) and varies from font to font.  The difference is obvious in print, although I think the improved character spacing can be seen even on a screen.  The demo shows three fonts: Cochineal, an OTF free font; Kings Caslon, a TTF commercial font; and Warnock Pro, an OTF commercial font.  The difference is most apparent on the Kings Caslon sample.  Warnock is much less affected, but it takes on an almost buttery smooth color with the correct kerning.  Fonts that do not use OFF-style kerning are not affected--the kerning is already correct for them.  To run the test files, they  will need to be modified to reflect each installation and to use the desired fonts.

The code changes are minor, just a matter of reading the table from the font file and hooking up the wires.  This zip file contains the full (modified) otf.c file and a git diff vs. the previous version (from 2015), showing the changed lines.  The current otf.c on Github is newer and has been updated to appease Clang, so applying the diff file to the current otf.c without human interpretation will undo those changes.

[Source.zip](https://github.com/n-t-roff/heirloom-doctools/files/14656287/Source.zip)

I've been unable to get the n-t-roff project to build successfully on my current machine.  The compiler exits with several linker errors of the form `/usr/bin/ld: eqnbox.c:(.text+0xe8): undefined reference to 'yyval'`.  If this is (still) the old Bison-related issue, it might be worth while to change the build process so that it uses only known-good files, and the yacc/lex/sed-related parts are not invoked--they would be relegated for use only in a development context.

--- a/troff/troff.d/otf.c
+++ b/troff/troff.d/otf.c
@@ -2815,10 +2815,43 @@ get_PairPosFormat1(int o)
 	free_cov(cp);
 }
 
+static int
+get_left_class(struct class *cp, int gid)
+{
+	int Class = 0;
+
+	if (cp->ClassFormat == 1) {
+		int g = gid - cp->StartGlyph;
+
+		if (g >= 0 && g < cp->GlyphCount)
+			Class = pbe16(&contents[cp->offset + 6 + 2 * g]);
+		return Class;
+	}
+	if (cp->ClassFormat == 2) {
+		int RangeRecord_offset, Start, End, i;
+
+		for (i = 0; i < cp->ClassRangeCount; i++) {
+			RangeRecord_offset = cp->offset + 4 + 6 * i;
+			Start = _pbe16(&contents[RangeRecord_offset]);
+			End = _pbe16(&contents[RangeRecord_offset + 2]);
+			if (gid < Start)
+				break;
+			if (gid >= Start && gid <= End) {
+				Class = _pbe16(&contents[RangeRecord_offset + 4]);
+				break;
+			}
+		}
+		return Class;
+	}
+	return -1;
+}
+
 static void
 get_PairPosFormat2(int o)
 {
 	struct class	*c1, *c2;
+	struct cov	*cp;
+	int	Coverage;
 	int	ValueFormat1, ValueFormat2;
 	int	ClassDef1, ClassDef2;
 	int	Class1Count, Class2Count;
@@ -2828,6 +2861,9 @@ get_PairPosFormat2(int o)
 	int	i, n, a;
 	int	x;
 
+	Coverage = o + pbe16(&contents[o+2]);
+	if ((cp = open_cov(Coverage)) == NULL)
+		return;
 	ValueFormat1 = pbe16(&contents[o+4]);
 	ValueFormat2 = pbe16(&contents[o+6]);
 	ClassDef1 = o + pbe16(&contents[o+8]);
@@ -2850,7 +2886,8 @@ get_PairPosFormat2(int o)
 				v2[n] = v;
 				n++;
 			}
-			while (get_class(c1, &g, &v), g >= 0) {
+			while ((g = get_cov(cp)) >= 0) {
+				v = get_left_class(c1, g);
 				if (v < 0 || v >= Class1Count)
 					continue;
 				for (i = 0; i < n; i++) {
@@ -2865,6 +2902,7 @@ get_PairPosFormat2(int o)
 		}
 		free_class(c1);
 	}
+	free_cov(cp);
 	free(g2);
 	free(v2);
 }
@@ -3129,10 +3167,11 @@ get_lookup(int o, int type, const char *name,
 		y = pbe16(&contents[LookupList+2+2*x]);
 		if ((t = pbe16(&contents[LookupList+y])) == type || type < 0) {
 			SubTableCount = pbe16(&contents[LookupList+y+4]);
-			for (j = 0; j < SubTableCount; j++)
+			for (j = 0; j < SubTableCount; j++) {
 				func(t, LookupList+y +
 					pbe16(&contents[LookupList+y+6+2*j]),
 					name);
+			}
 		}
 	}
 }

